### PR TITLE
Preserve channels_last in InstanceNorm and fix the channels-last group norm reductions at D == 1

### PR DESCRIPTION
**Pull request by dmenig**
Addresses the `InstanceNorm3d` item of #59168. Deliberately not a `Fixes:` — the other items in that issue stay open.

## The layout problem

`instance_norm` folds `N` into `C` and calls `batch_norm` on a `(1, N*C, ...)` view. That view cannot exist for a channels_last tensor: memory ordered `N,D,H,W,C` cannot be reindexed so `N*C` is innermost. The `input.contiguous()` that follows means every `InstanceNorm` in a channels_last model converts its input and returns a contiguous output the next convolution converts back. The cost of this was measured on nnU-Net in https://github.com/pytorch/pytorch/issues/59168#issuecomment-5339628979.

Instance norm is group norm with one group per channel, and `native_group_norm` already has channels_last kernels, so this dispatches there when the input is channels_last and there are no running stats to update — the default for the `nn.InstanceNorm*d` modules. `group_norm_memory_format` is shared out of `group_norm.cpp` so the fast path triggers exactly on the backends whose group norm kernels keep the layout.

## Dispatching alone made it slower

This is the first caller to reach the channels-last group norm kernels with one group per channel. Those kernels parallelize over groups and reduce `HxW` inside a block — fine for group norm's usual `G=32` over many channels, degenerate when `G == C` and so `D == 1`. At the reported shape the moments kernel launched 16 blocks and `ComputeInternalGradientsChannelsLast` launched **4**, the latter sustaining 21.6 GB/s on a card that does 561 GB/s.

The fix is a reduction that keeps `blockDim.x` on the channel axis, so a warp's reads stay contiguous, while `blockDim.y`/`gridDim.y` split `HxW` so the grid stays wide regardless of `C`. Partials are staged in global memory and combined by the last block to arrive. That is the strategy `batch_norm`'s channels-last statistics kernels already use — and the reason the *contiguous* instance norm path is fast today. The generic parts move out of `Normalization.cuh` into a new `channels_last_reduce.cuh` so `group_norm_kernel.cu` can use them without pulling in the batch norm kernels.

Where it is used differs by direction, because the two produce different quantities:

- **Backward** already reduces to per-`(n, c)` sums, so the new kernel replaces both channels-last variants outright; the fallback kernel and the heuristic choosing between them are deleted.
- **Forward** reduces to per-`(n, g)` moments. At `D == 1` those are the same thing, so it writes group statistics directly. At `D > 1` it would have to compute `C` moments and merge down to `G` — redundant work, and measured slower on `2,640,32,32,32` (`D=20`) even before counting the merge pass. So `D > 1` keeps the existing kernels unchanged, and only the `D == 1` arm of the old `(D == 1 || D == 2)` condition moves.

Two alternatives were tried and rejected. An occupancy threshold steering the existing kernels recovered the reported shape but regressed `N=8` by 1.5x and `N=16` by 1.9x, because the two old kernels differ in coalescing as well as grid width and no threshold separates those cases. Using the new reduction for every `D` regressed ordinary GroupNorm forward by up to 26%, for the redundancy above.

## Mixed dtype

The CUDA kernels read affine params and wrote saved stats and `dgamma`/`dbeta` in the input's scalar type, so fp32 params over a reduced-precision input (autocast) were downcast. The CPU kernel already used a separate param type; the same `PT` is now threaded through the CUDA kernels, and `dgamma`/`dbeta` are allocated from that param type rather than from `gamma.options()` — those disagree when `gamma` is undefined and only `beta` is fp32, which previously threw from the kernel on both CUDA and CPU.

## Behaviour change worth flagging

`eps` used to be cast to the input dtype. `1e-5` is below half's smallest normal, so **fp16 GroupNorm was effectively running with a subnormal or zero epsilon**. It is now kept at accumulation precision. That is a fix, but it changes fp16 `GroupNorm` results for existing callers on the *contiguous* path too, not only `InstanceNorm`.

## Numbers

TITAN RTX (sm_75, 72 SMs), measured as CUDA kernel time rather than wall clock, with the contiguous path measured alongside as a control that the two builds are comparable (it matches within 0.5% on every row).

`InstanceNorm3d` at the reported shape `(2, 32, 64, 64, 64)`, channels_last_3d:

| | before | after |
|---|---|---|
| moments | 1.640 ms | 0.171 ms |
| internal gradients | 6.208 ms | 0.241 ms |
| whole layer | 7.393 ms | **1.598 ms** (contiguous: 1.685 ms) |

Ordinary GroupNorm, channels-last, forward+backward:

| shape | before | after |
|---|---|---|
| `(16,256,32,32)` G=32 | 0.449 | 0.450 |
| `(8,512,16,16)` G=32 | 0.163 | 0.131 |
| `(2,128,64,64)` G=32 | 0.220 | 0.133 |
| `(32,64,56,56)` G=32 | 0.661 | 0.642 |
| `(1,320,64,64)` G=32 | 0.276 | 0.187 |
| `(2,128,16,32,32)` G=32 | 0.803 | 0.463 |
| `(4,256,8,32,32)` G=32 | 0.979 | 0.832 |

Forward and backward separately on five further configurations, since a backward win hides a forward regression in the combined number:

| shape | fwd | fwd+bwd |
|---|---|---|
| `8,64,64,56,56` fp16 | 0.0481 → 0.0481 | 0.1847 → 0.1412 |
| `2,640,32,32,32` fp32 | 0.0518 → 0.0518 | 0.1785 → 0.1616 |
| `8,32,8,64,64` fp16 | 0.0284 → 0.0282 | 0.1396 → 0.0918 |
| `8,256,256,28,28` fp16 | 0.0539 → 0.0441 | 0.1469 → 0.1419 |
| `4,128,32,112,112` fp16 | 0.1353 → 0.1352 | 0.6618 → 0.4352 |

**The end-to-end script is not conclusive on this card.** With the norm layer, channels_last_3d goes from +26.5% to +2.4% against contiguous — but the same script with `nn.Identity` in place of the norm still shows +7.7%, because Turing has no NHWC 3D convolution path and cuDNN falls back to `volta_scudnn` NCHW kernels. The norm layer's own contribution moves from roughly +19 points to roughly −5. Re-running that script on Ampere or newer would give the number this actually needs.

## Review order

`channels_last_reduce.cuh` (code motion) → the matching deletions in `Normalization.cuh` → `group_norm_kernel.cu`, where `RowwiseMomentsChannelsLastUnitDCUDAKernel` and `ComputeInternalGradientsChannelsLastCUDAKernel` are the new reductions and the rest is `PT` threading → `Normalization.cpp` with `group_norm.{cpp,h}` → tests.

## Testing

```
python test/test_nn.py -k instancenorm
python test/test_nn.py -k InstanceNorm
python test/test_nn.py -k GroupNorm
python test/test_nn.py -k group_norm
python test/test_modules.py -k InstanceNorm
python test/test_modules.py -k GroupNorm
python test/test_ops.py -k native_group_norm
```

The pre-existing norm tests all reduce a small `HxW`, which keeps the new kernels on their single-block path, so four tests cover what they miss:

- `test_group_norm_channels_last_staged_reduction` — an `HxW` large enough to split the reduction across blocks and exercise the staging and semaphore combine.
- `test_group_norm_channels_last_large_batch` — `N` above 65535, the `gridDim.z` limit the kernels walk the batch to stay under.
- `test_group_norm_channels_last_many_channels` — a channel count past the point the launch sizing saturates.
- `test_group_norm_mixed_dtype_bias_only` — a fp32 bias with no weight.

## Known limits

- **ROCm is untested.** This uses `__threadfence()` where `batch_norm`'s channels-last kernels deliberately use committed stores instead; correct but possibly slower there.
- Measured on sm_75 only.
- A **dynamic channel count** (`torch.compile(dynamic=True)`, `mark_dynamic` on the channel dim, or export) keeps the batch_norm path and so returns a contiguous output where eager returns channels_last. The comment at the guard records this. Specializing instead would trade the layout difference for a guard on a dimension that is almost always static.
- Routing channels_last instance norm through group norm means argument checking is group norm's, so a few malformed calls report a different message than the contiguous path does, and a degenerate `HxW == 1` lands within rounding of zero rather than exactly zero.

---

Authored with the assistance of Claude Code. The CUDA here deserves close reading rather than trust.

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude review this PR

**Comment by Skylion007** [comment]
Why are we implementing CUDA channels-last InstanceNorm through GroupNorm rather than dispatching to cuDNN’s native NHWC InstanceNorm? Have we benchmarked the two, particularly given the GroupNorm kernel’s low occupancy here? I think it might need a special fast path for the cudnn dispatch, but it should be able to handle that.

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -740,9 +742,39 @@ Tensor instance_norm(
...
+  // output, forcing a layout conversion around every InstanceNorm of a
+  // channels_last model. native_group_norm has channels_last kernels, so prefer
+  // it when there are no running stats to update.
+  const auto memory_format = group_norm_memory_format(input);
+  const auto num_groups = c.maybe_as_int();
+  if ((memory_format == MemoryFormat::ChannelsLast ||
+       memory_format == MemoryFormat::ChannelsLast3d) &&
+      use_input_stats && !running_mean.defined() && !running_var.defined() &&
+      num_groups.has_value()) {
+    // native_group_norm's CUDA kernels read weight/bias with the input's
```
Comment (Skylion007) [inline aten/src/ATen/native/Normalization.cpp]: Can we just fix native group norms kernel? The channel last path is new so this feels like a bug?

Comment (eqy) [reply aten/src/ATen/native/Normalization.cpp]: sure, let's see if this introduces a substantial performance penalty, if not we can change both paths to avoid downcast

Comment (Skylion007) [reply aten/src/ATen/native/Normalization.cpp]: There is also a cudnn fast path for channels_last instance norm we may just want to dispatch there...

**Comment by dmenig** [comment]
For Claude : 
>Addressed: finding 1 by keeping fp32 params/stats/grads in the kernel (no dtype gating, no downcast); 3 by dropping the redundant use_input_stats; 4 with a comment noting dynamic C opts out; 6 by rewriting the module_db comment to state that its inputs mix track_running_stats and the layout coverage lives in test_instancenorm_channels_last; 7 with a new mixed-dtype test asserting mean/rstd and affine grads stay fp32; 8 by skipping only mps/xpu instead of allow-listing. Finding 5 (two numerical paths by stride) is acknowledged as inherent to the layout routing.

File: aten/src/ATen/native/Normalization.cpp (hunk 1 above)
Comment (dmenig) [reply aten/src/ATen/native/Normalization.cpp]: Done. 
>Fixed the kernel rather than working around it. native_group_norm's CUDA kernels now take a separate param type PT (threaded through both forward and backward), so fp32 affine params, the saved mean/rstd, and dgamma/dbeta are kept in fp32 over a reduced-precision input, matching what the CPU kernel already did via param_t. The .to(input_dtype) cast in instance_norm is removed, so both the group_norm and instance_norm paths avoid the downcast. eps is also no longer narrowed to the input dtype. 

I'm in the process of measuring the perf delta @eqy  asked about. On the cuDNN-fast-path idea: leaving that out of scope for this PR un less you wanna

Comment (dmenig) [reply aten/src/ATen/native/Normalization.cpp]: Also @Skylion007  I'm being told by my LLM
>GroupNorm is the mathematically exact, already-in-tree, layout-native, dependency-free implementation; cuDNN has no instance-norm op and its batch-norm route needs the N→C fold that forces the very copy this PR eliminates. A cuDNN fast path is worth exploring only if the benchmark shows the group_norm kernel's occupancy costs more than the copies it saves

If you disagree, tell me so and I'll go down this path instead.

Comment (dmenig) [reply aten/src/ATen/native/Normalization.cpp]: The current solution isn't better
>the copies go away, but instance norm's G == C degenerates the channels-last group-norm kernels' parallelization to 4 blocks, so the norm itself becomes the new bottleneck.

it's actually even slower than main on my Titan RTX. I'm gonna rework this. Maybe open a new PR. i'll see where I land.

Comment (dmenig) [reply aten/src/ATen/native/Normalization.cpp]: Ok. I updated the PR.
>InstanceNorm3d now preserves channels_last_3d instead of round-tripping through contiguous(): it dispatches to native_group_norm, and the channels-last group norm CUDA reductions are fixed, since they collapsed to a handful of blocks when there is one group per channel (at the shape reported above the norm layer goes 7.39ms -> 1.60ms and is now faster than the contiguous path). The residual gap left in the repro on my Turing card is conv3d rather than the norm, so a re-run on Ampere or newer would be the number that actually settles it — PR is #194616.

Disclaimer guys. Contrary to what is said in AI_POLICY.md, I didn't proof read this. It's really not my programming language. I use an iterative review process and didn't stop until it saw no problem.

Comment (dmenig) [reply aten/src/ATen/native/Normalization.cpp]: Also @Skylion007 
image
